ocfweb/lab_reservations/reserve.py

```diff
@@ -0,0 +1,169 @@
+
+
+class NewReservationRequest(namedtuple(
+    'NewReservationRequest', [
+        'real_name',
+        'contact_email',
+        'group',
+        'reason',
+        'date',
+        'starttime',
+        'endtime',
+    ],
+)):
+    __slots__ = ()
+
+    def to_dict(self):
+        return self._asdict()
+
+
+class RequestForm(forms.Form):
+
+    error_css_class = 'error'
+    required_css_class = 'required'
+
+    real_name = forms.CharField(
+        label='Full Name',
+        widget=forms.TextInput(attrs={'placeholder': 'Oski Bear'}),
+        min_length=3,
+    )
+
+    contact_email = forms.EmailField(
+        label='Contact email',
+        validators=[wrap_validator(ocflib.misc.validators.valid_email)],
+        widget=forms.EmailInput(attrs={'placeholder': 'oski@berkeley.edu'}),
+    )
+
+    verify_contact_email = forms.EmailField(
+        label='Confirm contact email',
+        validators=[wrap_validator(ocflib.misc.validators.valid_email)],
+        widget=forms.EmailInput(attrs={'placeholder': 'oski@berkeley.edu'}),
+    )
+
+    group = forms.CharField(
+        label='Group',
+        widget=forms.TextInput(attrs={'placeholder': 'Open Computing Facility'}),
+        min_length=3,
+    )
+
+    reason = forms.CharField(
+        label='Reason for reservation',
+        widget=forms.Textarea(attrs={'placeholder': ''}),
+    )
+
+    date = forms.DateField(
+        label='Date of reservation (yyyy-mm-dd)',
+        widget=forms.DateInput(attrs={'placeholder': datetime.now().strftime('%Y-%m-%d')}),
+    )
+
+    starttime = forms.TimeField(
+        label='Starting time of reservation (24 hour, i.e. 20:00)',
+    )
+
+    endtime = forms.TimeField(
+        label='Ending time of reservation (24 hour, i.e. 22:00)',
+    )
+
+    disclaimer_agreement = forms.BooleanField(
+        label='I agree with the above statement.',
+        error_messages={
+            'required': 'You must agree to our policies.',
+        },
+    )
+
+    def clean_verify_(self):
+        email = self.cleaned_data.get('contact_email')
+        verify_contact_email = self.cleaned_data.get('verify_contact_email')
+
+        if email != verify_contact_email:
+            raise forms.ValidationError("Your emails don't match.")
+        return verify_contact_email
+
+
+@calnet_required
+def request_reservation(request):
+    calnet_uid = request.session['calnet_uid']
+    status = 'new_request'
+
+    # ensure we can even find them in university LDAP
+    # (alumni etc. might not be readable in LDAP but can still auth via CalNet)
+    if not user_attrs_ucb(calnet_uid):
+        return render(
+            request,
+            'lab_reservations/lab_reservations/cant-find-in-ldap.html',
+            {
+                'calnet_uid': calnet_uid,
+                'title': 'Unable to read account information',
+            },
+        )
+
+    if request.method == 'POST':
+        form = RequestForm(request.POST)
+        if form.is_valid():
```

Comment:
So if the form isn't valid here, doesn't this not re-render the form because the `else` is indented to be only on a get request? What happens if the form is invalid but made a POST request?

Reply:
If the form is invalid but makes a POST request, we just return the same page with everything still filled out (since the return on line 138 is not in any of these conditionals). Looks something like this: 
![image](https://user-images.githubusercontent.com/22068978/38156606-12b2172c-3434-11e8-8eb2-8bcac302d576.png)

Reply:
Ah yeah, I missed that section, thanks.